(#59) (#83) (#89) 운동기록캘린더 개발, 소셜회원 통합에따른 재로그인 안내로직 개발, 각종 약관 및 구글폼 연결

ldg-96 opened the pull request:
# 📝작업 내용

## (#59) 운동기록 캘린더 개발
- 캘린더의 UI는 직접 개발하고 `date-fns/locale`  `@use-gesture/react` `@react-spring/web` 3가지 라이브러리를 활용, 상호작용을 관리하여 구현하였습니다.

## (#83) 소셜회원 통합에 따른 재로그인 안내 로직 개발
- 구글아이디로 회원가입 후 해당 구글이메일의 카카오톡으로 로그인 시도 시 두개의 아이디가 통합되도록 백엔드에서 구현하였습니다. 프론트쪽에서는 아이디가 통합될 시 로그인 후 콜백페이지에서 url 쿼리 파라미터에 통합되었다는 스트링이 포함되게 되는대 이를 통하여 사용자에게 아이디가 통합되었다는 토스트 메세지를 띄워준 후 재로그인을 하도록 유도하였습니다.

## (#89) 각종 약관 및 구글 폼 연결
- 레포지토리에 legal-docs를 추가하였며 구글액션즈 워크플로우와 내부 스크립트를 통해서 CI/CD할때 .md파일로 저장된 각종 약관 문서들을 html로 변경하여 s3에 배포되도록 하였습니다.
- 해당 html을 웹뷰나 iframe으로 띄워야 하는데 CSP정책에 걸려 https://www.undabang.store/legal/* 경로에 대해서는 csp를 완화하여 iframe으로 해당 html을 표시할 수 있도록 하였습니다.

ldg-96 committed:
2025-06-14

ldg-96 committed:
2025-06-17

ldg-96 committed:
2025-06-17

ldg-96 committed:
2025-06-18

ldg-96 committed:
Merge branch 'develop' of github.com:projects200/front-server into feature/exercise-calendar

# Conflicts:
#	src/app/exercise/page.tsx
#	src/hooks/useExerciseApi.ts
#	temp.ts

ldg-96 committed:
Merge branch 'develop' of github.com:projects200/front-server into feature/exercise-calendar

ldg-96 committed:
Merge branch 'develop' of github.com:projects200/front-server into feature/exercise-calendar

# Conflicts:
#	temp.ts

ldg-96 committed:
캘린더 스와이프 적용 전 임시저장 커밋

ldg-96 committed:
(#59) (#83) (#89) 운동기록캘린더 개발, 소셜회원 통합에따른 재로그인 안내로직 개발, 각종 약관 및 구글폼 연결

ldg-96 committed:
(#59) (#83) (#89) 운동기록캘린더 개발, 소셜회원 통합에따른 재로그인 안내로직 개발, 각종 약관 및 구글폼 연결

ldg-96 committed:
(#59) (#83) (#89) 운동기록캘린더 개발, 소셜회원 통합에따른 재로그인 안내로직 개발, 각종 약관 및 구글폼 연결

SimYunSup commented on src/app/exercise/_components/monthView.tsx:
```diff
@@ -0,0 +1,78 @@
+  const router = useRouter()
+  const [notCacheData, setNotCacheData] = useState(false)
+
+  // counts값이 변경될때마다 캐시된 데이터인지 여부를 판단하여 스탬프 아이콘에 fadeIn 효과를 줄지 즉시 표시해줄지 판단합니다.
+  useEffect(() => {
+    const isEmpty = Object.keys(counts).length === 0
+    if (!notCacheData && isEmpty) setNotCacheData(true)
+  }, [counts])
```
**suggestion(non-blocking):** counts값이 변경될 때마다 fadeIn 효과를 줄지 판단하는거면 `useMemo`를 쓰는게 어떨까요?

SimYunSup commented on src/app/exercise/_components/monthView.tsx:
```diff
@@ -0,0 +1,78 @@
+  today,
+  counts,
+}: Props) {
+  const router = useRouter()
```
**suggestion(blocking)**: `<MonthView>`를 만들고 싶으시다면 router를 내부에 두는 것이 아닌 `onClick`같은 방식으로 넘겨주는 것이 재사용성에 좋지 않을까요?

SimYunSup commented on src/hooks/useExerciseApi.ts:
```diff
@@ -88,9 +151,15 @@ export const usePatchExercise = (exerciseId: number) =>
 export const useDeleteExercise = (exerciseId: number) =>
   useApiMutation<null, null>(
-    ['exercise/delete', exerciseId],
+    ['exercise/delete'],
```
**question(blocking)**: 이 id와 밑의 delete만 key에서 지운 이유가 있을까요?

SimYunSup replied:
Mutation의 결과값은 캐싱이 안됩니다. mutation의 key가 존재하는 이유는 exerciseId의 데이터 값을 쉽게 revalidate하기 위해 있는 것인데, 지우기보다 모두 넣는게 도움이 되는 것 같은데 어떻게 생각하시나요?

ldg-96 commented on src/app/auth/callback/_components/callbackClient.tsx:
```diff
@@ -0,0 +1,64 @@
+  const [errorDescription] = useQueryState('error_description')
+  const hasBeenHandled = useRef(false)
+
+  const { data: registeredData } = useReadRegistered(
+    auth.isAuthenticated ? ['auth/isRegistered'] : null,
+  )
```
SWR의 조건부 데이터 가져오기 기능을 활용하여 보았습니다. 소셜로그인(코그니토) 후 콜백 페이지로 리디렉션 되엇을때 인증상태가 즉시 true가 되는 것이 아니라 약간의 로딩시간이 있습니다. 따라서 callback페이지 진입 초기시점에는 API호출에 필요한 인증 토큰이 아직 준비되지 않은 상태입니다. 

https://swr.vercel.app/docs/conditional-fetching

SWR의 조건부 패칭 기능을 활용하여 아직 토큰이 준비되지 않은 상황에서는 null값을 넣어 fetcher 함수를 실행하지 않도록 하였고 로딩이 끝나고 isAuthenticated값이 true가되어 토큰이 준비되었을때 key값을 넣어 api호출을 하도록 하였습니다.

ldg-96 committed:
(#59) (#83) (#89) 운동기록캘린더 개발, 소셜회원 통합에따른 재로그인 안내로직 개발, 각종 약관 및 구글폼 연결
수정내용: weeks생성부분 exerciseCalendar에서 monthView로 이동

ldg-96 committed:
(#59) (#83) (#89) 운동기록캘린더 개발, 소셜회원 통합에따른 재로그인 안내로직 개발, 각종 약관 및 구글폼 연결
수정내용: usePatchExercise(운동기록 수정) 훅의 key값에서 파라미터(exerciseId) 제거

ldg-96 committed:
(#59) (#83) (#89) 운동기록캘린더 개발, 소셜회원 통합에따른 재로그인 안내로직 개발, 각종 약관 및 구글폼 연결
수정내용: monthView의 재사용성을 위하여 router를 exerciseCalendar로 이동, 클릭시 로직을 exerciseCalendar에서 onDateClick함수를 만들어 monthView에 전달하도록 수정

SimYunSup commented on src/app/auth/callback/page.tsx:
```diff
@@ -1,53 +1,10 @@
-
-  return null
+  return (
+    <Suspense>
+      <CallbackClient />
+    </Suspense>
```
최상단에서 'use client'를 지워서가 아닐까요.

ldg-96 replied:
```
'use client'

import CallbackClient from './_components/callbackClient'

export default function CallbackPage() {
  return (
    <>
      <CallbackClient />
    </>
  )
}
```
아닙니다 이와같이 최상단에 'use client'를 적용하여도 
` ⨯ useSearchParams() should be wrapped in a suspense boundary at page "/auth/callback". Read more: https://nextjs.org/docs/messages/missing-suspense-with-csr-bailout`
`Error occurred prerendering page "/auth/callback". Read more: https://nextjs.org/docs/messages/prerender-error
``Export encountered an error on /auth/callback/page: /auth/callback, exiting the build.`
이와같은 useSearchParams() 쿼리파라미터 사용에 관한 에러가 발생하게됩니다.

SimYunSup replied:
![스크린샷 2025-06-27 오후 2 44 31](https://github.com/user-attachments/assets/0b683e5a-a929-40d8-b4e1-43628d095117)
minimal로 바꾸면 아주 잘 되는데 다시 디버깅해보시죠.

ldg-96 replied:
```
'use client'

import { useQueryState } from "nuqs"

export default function Temp() {

  const [date, setDate] = useQueryState('date')

  return <div></div>
}
```
위와같이 최대한 미니멀하게 코드를 작성해서 빌드를 해보았는데도 
useSearchParams() should be wrapped in a suspense boundary at page "/temp".
해당 빌드오류가 발생하였습니다.

콜백페이지와 리스트페이지의 차이점에 대해서 생각해본 결과 리스트 페이지는 상단에 layout.tsx에서 AuthGuard컴포넌트로 감싸져 있다는 점이 있었습니다.

```
export default function AuthGuard({ children }: { children: React.ReactNode }) {
  const auth = useAuth()
  const router = useRouter()
  const pathname = usePathname()

  useEffect(() => {
    // 인증 로딩중
    if (auth.isLoading) return

    // 인증안된유저
    if (!auth.isAuthenticated) {
      router.replace(SITE_MAP.LOGIN)
      return
    }

    // 인증완료, 가입유저가 "/"에 있는경우
    if (auth.isAuthenticated && pathname === '/') {
      router.replace(SITE_MAP.EXERCISE)
      return
    }
  }, [auth.isLoading, auth.isAuthenticated])

  if (auth.isLoading) return null

  return <>{children}</>
}
```
AuthGuard는 위와같은 코드로 이루어져 있습니다. AuthGuard로 감싸져있는 페이지인 리스트 페이지를 렌더링할경우 다음과 같은 순서대로 진행이 됩니다.

build시 /list 페이지를 정적으로 생성 시도 -> layout이 먼저 렌더링되므로  AuthGuard실행 -> useAuth()호출되며 isLoading상태가 true이기 때문에 retun null -> 따라서 list페이지는 빌드시 렌더링 조차 되지 않음.

따라서 AuthGuard가 문제를 해결했던 원리를 콜백페이지에 적용하여 구현해보았습니다. 컴포넌트가 브라우저에 마운트되기 전까지는 문제가 되는 훅을 호출하는 로직을 실행하지 않도록 아래와 같이 코드를 작성해 보았습니다.

ldg-96 committed:
(#59) (#83) (#89) Callback페이지에서 훅을 사용할때 key를 전달하지 않고shouldFetch 인자를 받도록 수정

ldg-96 committed:
(#59) (#83) (#89) CRUD중 C를 제외한 Mutation의 Key값에 파라미터값 부여

ldg-96 committed:
(#59) (#83) (#89) useMemo를 사용하여 fadeIn 애니메이션 여부 결정

ldg-96 committed:
빌드오류 해결

ldg-96 committed:
회원가입, 운동기록 QA

ldg-96 commented:
(#82) 운동기록 QA
- 로그인 통합메세지 줄바꿈
- 연도선택 100년 -> 광복이후만 가능
- 생년월일 초기값 YYYY/MM/DD
- 성별 초기값 없음

위 목록에 대해서 추가적으로 코드변경이 있었습니다. 자잘한부분이고 회원가입에 대해서는 추후 시간날때 tanstack과 zod를 활용한 코드로 변경할 예정이기에 별도로 리뷰해주지 않으셔도 될것 같습니다!

ldg-96 committed:
회원가입, 운동기록 QA

ldg-96 committed:
회원가입, 운동기록 QA

ldg-96 committed:
(#59) (#83) (#89) callback 페이지 useSearchParams() 빌드오류 해결

ldg-96 committed:
esLinst no-unused-vars 제거
